### PR TITLE
Fix site-wide 404 & Pass along HTTP status info to error handler

### PR DESCRIPTION
**Pull request by confused-Techie**
With the brand new logging we have setup, I was able to find a bug occurring in prod. While the error handler expected an object, it was only being given a `404` status code number during the site wide 404 error. Which would cause it to crash.

This PR fixes this, ensuring that site wide 404's are caught properly by the generic error handler. 
Additionally, I've gone ahead and added some logic that lets other errors, such as those when displaying an individual package page, can bubble up the HTTP status code they received to the error handler, providing the HTTP error code the user will actually receive. Which can help make the logs easier to inspect, and keeps things much more accurate

### CODE REVIEW

**Comment by DeeDeeG** [comment]
>  Additionally, I've gone ahead and added some logic that lets other errors, such as those when displaying an individual package page, can bubble up the HTTP status code they received to the error handler, providing the HTTP error code the user will actually receive. Which can help make the logs easier to inspect, and keeps things much more accurate

Big 👍 on this! Reviewing the diff now...

File: src/main.js [hunk 1]
```diff
@@ -69,7 +71,19 @@ app.get("/logout", async (req, res) => {
...
+      timecop: false,
+      page: {
+        name: "PPR Error Page",
+        og_url: "https://web.pulsar-edit.dev/packages",
+        og_description: "The Pulsar Package Repository",
+        og_image: "https://web.pulsar-edit.dev/public/pulsar_name.svg",
+        og_image_type: "image/svg+xml"
+      },
+      status_to_display: 404
+    });
```
Comment (DeeDeeG) [inline src/main.js:86]: Is this intended to be indented 4 spaces, or should it be just 2 spaces indented?

File: src/handlers.js [hunk 2]
```diff
@@ -65,6 +65,27 @@ async function singlePackageListing(req, res, timecop) {
       og_image_height: 600,
     }});
   } catch(err) {
+    let status_to_display = false; // Since the status is ignored if not a number,
+    // we initialize as boolean to no-op in the case we don't find a proper status
+
+    const validStatus = (val, key) => {
+      if (val?.response?.[key] && typeof val.response[key] === "boolean" && val.response[key]) {
```
Comment (DeeDeeG) [inline src/handlers.js:72]: I would suggest streamlining this a tad.

A very shortened version of what I believe is an equivalent check:

```suggestion
      if (val?.response?.[key] === true) {
```

Or maybe this:

```suggestion
      if (typeof val?.response?.[key] === "boolean" && val.response[key]) {
```

^ if the convention of explicitly checking `typeof` is important, to signal to the reader "this is doing type checking," and to be less mistake prone with `true` (boolean) vs `"true"` (string) (kind of like "TypeScript but with less tooling," heh).

Third suggestion, why not:

```suggestion
      if (val?.response?.[key] && typeof val.response[key] === "boolean") {
```

File: src/handlers.js [hunk 3]
```diff
@@ -65,6 +65,27 @@ async function singlePackageListing(req, res, timecop) {
       og_image_height: 600,
     }});
   } catch(err) {
+    let status_to_display = false; // Since the status is ignored if not a number,
+    // we initialize as boolean to no-op in the case we don't find a proper status
+
+    const validStatus = (val, key) => {
```
Comment (DeeDeeG) [inline src/handlers.js:71]: Very optional suggestion, but this function name was not obvious to me what it did, I dunno if this is any better, but **how about calling this function  `validStatusIs`?** or `statusIs`?

Since it is checking not just validity overall (based on first argument matching some "set" of okay values), but rather specifically matching to the exact supplied status (second argument).

So it is more "matching" to a supplied okay value, rather than "checking validity" only. (Very minor point, go with what you like.)

File: src/utils.js [hunk 4]
```diff
@@ -23,7 +23,11 @@ const reg = require("./reg.js");
 
 async function displayError(req, res, details) {
   console.error(details);
-  res.status(500).render('error', details);
+  if (details?.status_to_display && typeof details.status_to_display === "number") {
```
Comment (DeeDeeG) [inline src/utils.js:26]:
```suggestion
  if (typeof details?.status_to_display === "number") {
```

This is an optional thing to streamline it, but I kind of think it is fine as-is to more clearly in plain English explain the two checks (is it there/truthy, and is it a number type?). So, honestly either syntax/style of expressing this being functionally equivalent, go for what style you like.

**Review by DeeDeeG — APPROVED** [review]
Looks good overall, I had some comments, they're all style things though, so I can approve it as-is basically.

Hope this isn't too thorough/small details.

Also, I am not sure which part of the backend this corresponds to, so I can't personally verify that this works correctly with the backend. Just reading the code from this diff only, and surrounding lines of code for context. With that level of info, it looke reasonable, and passing along specific error codes is _SO, SO GOOD_ from a debugging standpoint than "everything is code 500". Love having more error codes passed to the backend!!! ❤️

**Comment by DeeDeeG** [comment]
Checking if the indentation was as deep as intended is pretty straight-forward, and the rest are style things about some logic checks that I don't feel strongly about.

Overall 👍 regardless of these comments.

**Comment by confused-Techie** [comment]
@DeeDeeG Thanks a ton for giving this one a review, lots of fantastic feedback, especially about shortening those logic checks. I'll go ahead and address those properly tomorrow, just giving a nod to show I've seen it, thanks!

File: src/handlers.js (hunk 2 above)
Comment (DeeDeeG) [reply src/handlers.js:72]: Honestly, none of my suggestions here are more reasable while still explicitly checking `typeof ... === "boolean"`, so I think it's basically more readable as-is. I'll mark this as resolved.

**Comment by confused-Techie** [comment]
@DeeDeeG Thanks a ton for the feedback, sorry it took some time to get too.

I've gone ahead and addressed this feedback and updated the commit following all three of your well thought out suggestions. Thanks again and feel free to let me know what you think!

File: src/handlers.js [hunk 5]
```diff
@@ -65,6 +65,27 @@ async function singlePackageListing(req, res, timecop) {
       og_image_height: 600,
     }});
   } catch(err) {
+    let status_to_display = false; // Since the status is ignored if not a number,
+    // we initialize as boolean to no-op in the case we don't find a proper status
+
+    const validStatusIs = (val, key) => {
+      if (val?.response?.[key] && typeof val.response[key] === "boolean" && val.response[key]) {
```
Comment (DeeDeeG) [inline src/handlers.js:72]:
```suggestion
      if (typeof val?.response?.[key] === "boolean" && val.response[key]) {
```

On second thought, maybe this, and it does sort of match with `  if (typeof details?.status_to_display === "number") {` in the other file now.

**Review by DeeDeeG — APPROVED** [review]
Again, approved as-is, with optional style feedback.

I waffled a bit on the style of one of the lines (see suggestion below). THAT SAID, the PR is generally great, and I don't mean to be picky regarding feedback. So I emphasize that the feedback is once again optional.

Approve stands with/without the suggested code change. Feel free to merge with or without applying the suggested diff.

File: src/handlers.js (hunk 5 above)
Comment (confused-Techie) [reply src/handlers.js:72]: It's been a while since looking closely at this code.
But if memory serves, as of now, this should always be present, and there wouldn't be a case where it's fully undefined, but with that said, the idea is that in the future it may be totally possible for `val` to be totally undefined, in which case we might as well protect from it now, rather than let the problem sneak up on us.

So thanks for finding that flaw, and lets add this in so we are good to merge as soon as possible, thanks again for the awesome feedback on this one!

**Review by DeeDeeG — APPROVED** [review]

File: src/handlers.js (hunk 5 above)
Comment (DeeDeeG) [reply src/handlers.js]: INFORMATIONAL ONLY NOTE, NO ACTION NEEDED:

I am sad to report, after testing in the node REPL, that `typeof val?.response?.[key]` will error if `val` is not defined :/

`Uncaught ReferenceError: val is not defined`

So, this was not really an error safety thing after all, and one would have to `try/catch` if not throwing was important for `val` not being defined.

ONCE AGAIN< INFORMATIONAL ONLY< NO ACTION NEEDED, just thought I should share what I found in actual testing.
